File: core/common/uptest.cpp
#include "_string.h"
#include "stream.h"
#include "uptest.h"
#include "threading.h"
#include "socket.h"
#include "version2.h"
#include "xml.h"
#include "sstream.h"
#include <algorithm>
#ifdef WITH_RUST_CORE
#include "rustbridge.h"
#endif

// -------- class UptestServiceRegistry --------

std::pair<bool,std::string> UptestServiceRegistry::addURL(const std::string& url)
{
    std::lock_guard<std::recursive_mutex> cs(m_lock);

    URI uri(url);
#ifdef WITH_RUST_CORE
    // 加えてよいかは Rust (peercast-rs の src/uptest.rs) が決める。
    bool valid = uri.isValid();
    std::string scheme = valid ? uri.scheme() : "";
    std::vector<pcrs_bytes> existing;
    for (auto& e : m_providers)
        existing.push_back({ reinterpret_cast<const uint8_t*>(e.url.data()), e.url.size() });
    const char* error = pcrs_uptest_check_add_url(valid,
                                                  reinterpret_cast<const uint8_t*>(scheme.data()), scheme.size(),
                                                  reinterpret_cast<const uint8_t*>(url.data()), url.size(),
                                                  existing.data(), existing.size());
    if (error)
        return std::make_pair(false, error);
#else
    if (!uri.isValid())
        return std::make_pair(false, "invalid URL");
    if (uri.scheme() != "http")
        return std::make_pair(false, "unsupported protocol");
    if (0 < std::count_if(m_providers.begin(), m_providers.end(), [url](UptestEndpoint& e) { return e.url == url; }))
        return std::make_pair(false, "URL already exists");
#endif

    m_providers.push_back(UptestEndpoint(url));
    return std::make_pair(true, "");
}

bool UptestServiceRegistry::isIndexValid(int index) const
{
    return index >= 0 && index < m_providers.size();
}

std::pair<bool,std::string> UptestServiceRegistry::deleteByIndex(int index)
{
    std::lock_guard<std::recursive_mutex> cs(m_lock);

    if (!isIndexValid(index))
        return std::make_pair(false, "index out of range");

    m_providers.erase(m_providers.begin() + index);
    return std::make_pair(true, "");
}

std::pair<bool,std::string> UptestServiceRegistry::takeSpeedtest(int index)
{
    std::lock_guard<std::recursive_mutex> cs(m_lock);

    if (!isIndexValid(index))
        return std::make_pair(false, "index out of range");

    return m_providers[index].takeSpeedtest();
}

std::vector<std::string> UptestServiceRegistry::getURLs() const
{
    std::lock_guard<std::recursive_mutex> cs(m_lock);
    std::vector<std::string> res;
    for (auto provider : m_providers)
        res.push_back(provider.url);
    return res;
}

void UptestServiceRegistry::clear()
{
    std::lock_guard<std::recursive_mutex> cs(m_lock);
    m_providers.clear();
}

static std::string textStatus(int status)
{
#ifdef WITH_RUST_CORE
    const char* s = pcrs_uptest_text_status(status);
    if (!s)
        abort();
    return s;
#else
    switch (status)
    {
    case UptestEndpoint::kUntried:
        return "Untried";
    case UptestEndpoint::kSuccess:
        return "Success";
    case UptestEndpoint::kError:
        return "Error";
    default:
        abort();
    }
#endif
}

amf0::Value UptestServiceRegistry::getState()
{
    std::lock_guard<std::recursive_mutex> cs(m_lock);
    std::vector<amf0::Value> providers;

    for (size_t i = 0; i < m_providers.size(); i++)
        providers.push_back(
            {
                {"url", m_providers[i].url},
                {"status", textStatus(m_providers[i].status)},
                {"speed", m_providers[i].m_info.speed},
                {"over", m_providers[i].m_info.over},
                {"checkable", m_providers[i].m_info.checkable},
            });

    return amf0::Value::object(
        {
            {"providers", providers}
        });
}

void UptestServiceRegistry::update()
{
    std::lock_guard<std::recursive_mutex> cs(m_lock);

    for (auto& provider : m_providers)
    {
        if (provider.status != UptestEndpoint::kSuccess)
        {
            if (provider.isReady())
                provider.update();
            else
                LOG_TRACE("%s not ready to download", provider.url.c_str());
        }
    }
}

void UptestServiceRegistry::forceUpdate()
{
    std::lock_guard<std::recursive_mutex> cs(m_lock);

    for (auto& provider : m_providers)
        provider.update();
}

std::pair<bool,std::string> UptestServiceRegistry::getXML(int index, std::string& out) const
{
    std::lock_guard<std::recursive_mutex> cs(m_lock);

    if (!isIndexValid(index))
        return std::make_pair(false, "index out of range");

    if (m_providers[index].status != UptestEndpoint::kSuccess)
        return std::make_pair(false, "no XML");

    out = m_providers[index].m_xml;
    return std::make_pair(true, "");
}

// -------- class UptestEndpoint --------

bool UptestEndpoint::isReady()
{
#ifdef WITH_RUST_CORE
    return pcrs_uptest_is_ready(status, lastTriedAt, sys->getTime());
#else
    return (status == kUntried) || (sys->getTime() - lastTriedAt > kXmlTryInterval);
#endif
}

void UptestEndpoint::update()
{
    this->lastTriedAt = sys->getTime();
    LOG_DEBUG("Speedtest: %s", this->url.c_str());
    try
    {
        m_xml = download(this->url);
        m_info = readInfo(m_xml);
        this->status = kSuccess;
        LOG_INFO("Speedtest result: %s kbps", m_info.speed.c_str());
    }catch (std::exception& e)
    {
        LOG_ERROR("UptestEndpoint %s: %s", this->url.c_str(), e.what());
        this->status = kError;
    }
}

std::string UptestEndpoint::download(const std::string& url)
{
    URI uri(url);

    if (!uri.isValid())
        throw std::runtime_error("invalid URI");
    if (uri.scheme() != "http")
        throw std::runtime_error("unsupported protocol");

    Host host;
    host.fromStrName(uri.host().c_str(), uri.port());

    if (!host.ip)
        throw std::runtime_error("could not resolve host name");

    auto rsock = sys->createSocket();
    rsock->open(host);
    rsock->connect();

    HTTP http(*rsock);
    HTTPRequest req("GET", uri.path(), "HTTP/1.0",
                    {
                        { "Host", uri.host() },
                        { "Connection", "close" },
                        { "User-Agent", PCX_AGENT }
                    });
    HTTPResponse res = http.send(req);

    if (res.statusCode != 200)
        throw std::runtime_error(str::format("unexpected status code %d", res.statusCode));

    return res.body;
}

#define nonNull(p)                                                      \
    (((p) == nullptr) ?                                                 \
     throw std::runtime_error(                                          \
         str::format("non-null assertion failed on line %d in file %s", \
                     __LINE__, __FILE__))                               \
     : (p))

#ifdef WITH_RUST_CORE
// yp4g.xml は Rust (peercast-rs の src/uptest.rs) が読む。
UptestInfo UptestEndpoint::readInfo(const std::string& body)
{
    rustbridge::RustBuf out;
    switch (pcrs_uptest_read_info(reinterpret_cast<const uint8_t*>(body.data()), body.size(), out.out()))
    {
    case 0: break;
    case 3: throw StreamException("Tag too long");
    case 4: throw StreamException("Content too big");
    case 5: throw StreamException("Not XML document");
    case 6: throw StreamException("Unexpected end tag");
    case 7: throw StreamException("Too many attributes");
    case 8: throw StreamException("Bad tag value");
    default:
        throw std::runtime_error(str::format("non-null assertion failed on line %d in file %s",
                                             __LINE__, __FILE__));
    }

    UptestInfo info;
    std::string* fields[] = {
        &info.name, &info.ip, &info.port_open, &info.speed, &info.over, &info.checkable, &info.remain,
        &info.addr, &info.port, &info.object, &info.post_size, &info.limit, &info.interval, &info.enabled,
    };
    std::string s = out.str();
    size_t pos = 0;
    for (auto f : fields)
    {
        size_t end = s.find('\0', pos);
        *f = s.substr(pos, end - pos);
        pos = end + 1;
    }
    return info;
}
#else
UptestInfo UptestEndpoint::readInfo(const std::string& body)
{
    UptestInfo info;
    XML xml;
    StringStream mem(body);

    xml.read(mem);

    XML::Node* yp = nonNull(xml.findNode("yp"));
    info.name      = nonNull(yp->findAttr("name"));

    XML::Node* host = nonNull(xml.findNode("host"));
    info.ip        = nonNull(host->findAttr("ip"));
    info.port_open = nonNull(host->findAttr("port_open"));
    info.speed     = nonNull(host->findAttr("speed"));
    info.over      = nonNull(host->findAttr("over"));

    XML::Node* uptest = nonNull(xml.findNode("uptest"));
    info.checkable = nonNull(uptest->findAttr("checkable"));
    info.remain    = nonNull(uptest->findAttr("remain"));

    XML::Node* uptest_srv = nonNull(xml.findNode("uptest_srv"));
    info.addr      = nonNull(uptest_srv->findAttr("addr"));
    info.port      = nonNull(uptest_srv->findAttr("port"));
    info.object    = nonNull(uptest_srv->findAttr("object"));
    info.post_size = nonNull(uptest_srv->findAttr("post_size"));
    info.limit     = nonNull(uptest_srv->findAttr("limit"));
    info.interval  = nonNull(uptest_srv->findAttr("interval"));
    info.enabled   = nonNull(uptest_srv->findAttr("enabled"));

    return info;
}
#endif

std::string UptestEndpoint::generateRandomBytes(size_t size)
{
    std::string buf;
    peercast::Random r;

    for (size_t i = 0; i < size; ++i)
        buf += (char) (r.next() & 0xff);
    return buf;
}

HTTPResponse UptestEndpoint::postRandomData(URI uri, size_t size)
{
    Host host;
    host.fromStrName(uri.host().c_str(), uri.port());

    if (!host.ip)
        throw std::runtime_error("Could not resolve host name");

    auto rsock = sys->createSocket();
    rsock->open(host);
    rsock->connect();

    HTTP http(*rsock);
    HTTPRequest req("POST", uri.path(), "HTTP/1.0",
                    {
                        { "Host", uri.host() }, // this may not be necessary
                        { "Connection", "close" },
                        { "User-Agent", PCX_AGENT },
                        { "Content-Length", std::to_string(size) },
                        { "Content-Type", "application/octet-stream" }
                    });
    req.body = generateRandomBytes(size);
    return http.send(req);
}

std::pair<bool,std::string> UptestEndpoint::takeSpeedtest()
{
    // yp4g.xml をダウンロードして状態を得る。
    try
    {
        m_xml = download(this->url);
        m_info = readInfo(m_xml);
        LOG_DEBUG("Speedtest %s: checkable = %s", url.c_str(), m_info.checkable.c_str());
    }catch (std::exception& e)
    {
        LOG_ERROR("Speedtest %s: %s", url.c_str(), e.what());
        return std::make_pair(false, e.what());
    }

    if (m_info.checkable != "1")
    {
        LOG_ERROR("speedtest server unavailable: checkable != 1");
        return std::make_pair(false, "speedtest server unavailable");
    }else
    {
        // ランダムデータを POST する。
        const auto uptest_cgi = m_info.postURL();
        try
        {
            auto size = atoi(m_info.post_size.c_str()) * 1000;
            LOG_DEBUG("Posting %d bytes of random data to %s ...", (int) size, uptest_cgi.c_str());
            auto res = postRandomData(uptest_cgi, size);
            LOG_TRACE("... done");
            if (res.statusCode == 302)
                return std::make_pair(true, "");
            else
                return std::make_pair(false, str::format("unexpected status code %d", res.statusCode));
        }catch (std::exception& e)
        {
            LOG_ERROR("exception occurred while posting: %s", e.what());
            return std::make_pair(false, "exception occurred while posting");
        }
    }
}

// -------- class UptestInfo --------

std::string UptestInfo::postURL()
{
#ifdef WITH_RUST_CORE
    rustbridge::RustBuf url(pcrs_uptest_post_url(reinterpret_cast<const uint8_t*>(addr.data()), addr.size(),
                                                 reinterpret_cast<const uint8_t*>(port.data()), port.size(),
                                                 reinterpret_cast<const uint8_t*>(object.data()), object.size()));
    return url.str();
#else
    return str::format("http://%s:%s%s", addr.c_str(), port.c_str(), object.c_str());
#endif
}

File: peercast-rs/src/uptest.rs
//! 帯域測定 (core/common/uptest.cpp) のうち、通信しない部分。
//!
//! yp4g.xml の読み取り (`UptestEndpoint::readInfo`)、`UptestInfo::postURL`、`isReady`、
//! 状態の文字列、URL を加えてよいかの判断。ダウンロードと POST、ロックは C++ 側。

use crate::reader::SliceReader;
use crate::xml::{self, AttrError, Attributes};

/// `UptestInfo` の欄の数と並び
pub const FIELDS: [&str; 14] = [
    "name", "ip", "port_open", "speed", "over", "checkable", "remain", "addr", "port", "object", "post_size",
    "limit", "interval", "enabled",
];

/// `readInfo` がどのノードのどの属性を読むか (`FIELDS` の順)
const STEPS: [(&[u8], &[&[u8]]); 4] = [
    (b"yp", &[b"name"]),
    (b"host", &[b"ip", b"port_open", b"speed", b"over"]),
    (b"uptest", &[b"checkable", b"remain"]),
    (b"uptest_srv", &[b"addr", b"port", b"object", b"post_size", b"limit", b"interval", b"enabled"]),
];

/// `UptestInfo` (`FIELDS` の順)
pub type Info = [Vec<u8>; 14];

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ReadError {
    /// `XML::read` の失敗 (`Callback` にはならない)
    Xml(xml::Error),
    /// タグの属性が読めない ("Too many attributes" か "Bad tag value")
    Attr(AttrError),
    /// ノードか属性がない (C++ 版の `nonNull` の失敗)
    Null,
}

fn c_str(s: &[u8]) -> &[u8] {
    &s[..s.iter().position(|&c| c == 0).unwrap_or(s.len())]
}

/// `XML::Node` のうち、ここで使うもの
struct Node {
    attrs: Attributes,
    children: Vec<usize>,
    parent: Option<usize>,
}

impl Node {
    fn at(&self, pos: usize) -> &[u8] {
        c_str(self.attrs.data.get(pos..).unwrap_or(&[]))
    }

    /// `getName`
    fn name(&self) -> &[u8] {
        self.at(self.attrs.attrs[0].0)
    }

    /// `findAttr`: 名前の先頭が `name` と一致する (大文字小文字は問わない) 最初の属性の値
    fn find_attr(&self, name: &[u8]) -> Option<&[u8]> {
        self.attrs.attrs[1..].iter().find_map(|&(n, v)| {
            let an = self.at(n);
            (an.len() >= name.len() && an[..name.len()].eq_ignore_ascii_case(name)).then(|| self.at(v))
        })
    }
}

/// `XML` の木 (rustbridge.h の `XmlReader` が組み立てるものと同じ形)
#[derive(Default)]
struct Tree {
    nodes: Vec<Node>,
    root: Option<usize>,
    curr: Option<usize>,
    attr_error: Option<AttrError>,
}

/// `XML::Node::Node("%s", tag)` の書き込み先の大きさ
const NODE_TMP_LEN: usize = 8192;

impl xml::Builder for Tree {
    fn content(&mut self, _s: &[u8]) -> Result<(), ()> {
        Ok(())
    }

    fn start_tag(&mut self, s: &[u8], single: bool) -> Result<(), ()> {
        let tag = &s[..s.len().min(NODE_TMP_LEN - 1)];
        let attrs = xml::parse_attributes(tag).map_err(|e| self.attr_error = Some(e))?;
        let i = self.nodes.len();
        self.nodes.push(Node { attrs, children: Vec::new(), parent: self.curr });
        match self.curr {
            Some(p) => self.nodes[p].children.push(i),
            None => self.root = Some(i),
        }
        if !single {
            self.curr = Some(i);
        }
        Ok(())
    }

    fn end_tag(&mut self) -> Result<(), ()> {
        // Rust の xml::read は開いているノードがあるときだけ呼ぶ
        self.curr = self.curr.and_then(|c| self.nodes[c].parent);
        Ok(())
    }
}

impl Tree {
    /// `XML::findNode`: 木を先に親、次に子の順でたどり、名前が一致する (大文字小文字は
    /// 問わない) 最初のノード
    fn find_node(&self, name: &[u8]) -> Option<&Node> {
        let mut stack: Vec<usize> = self.root.into_iter().collect();
        while let Some(i) = stack.pop() {
            let n = &self.nodes[i];
            if n.name().eq_ignore_ascii_case(name) {
                return Some(n);
            }
            stack.extend(n.children.iter().rev());
        }
        None
    }
}

/// `UptestEndpoint::readInfo`
pub fn read_info(body: &[u8]) -> Result<Info, ReadError> {
    let mut tree = Tree::default();
    let mut r = SliceReader { data: body, pos: 0 };
    if let Err(e) = xml::read(&mut r, &mut tree) {
        return Err(match (e, tree.attr_error) {
            (xml::Error::Callback, Some(a)) => ReadError::Attr(a),
            _ => ReadError::Xml(e),
        });
    }

    let mut info: Info = Default::default();
    let mut k = 0;
    for (node, attrs) in STEPS {
        let n = tree.find_node(node).ok_or(ReadError::Null)?;
        for a in attrs {
            info[k] = n.find_attr(a).ok_or(ReadError::Null)?.to_vec();
            k += 1;
        }
    }
    Ok(info)
}

/// `UptestInfo::postURL`
pub fn post_url(addr: &[u8], port: &[u8], object: &[u8]) -> Vec<u8> {
    [&b"http://"[..], c_str(addr), b":", c_str(port), c_str(object)].concat()
}

/// `UptestEndpoint::Status`
pub const UNTRIED: i32 = 0;
pub const SUCCESS: i32 = 1;
pub const ERROR: i32 = 2;

/// `UptestEndpoint::kXmlTryInterval`
pub const XML_TRY_INTERVAL: u32 = 60;

/// `UptestEndpoint::isReady`
pub fn is_ready(status: i32, last_tried_at: u32, now: u32) -> bool {
    status == UNTRIED || now.wrapping_sub(last_tried_at) > XML_TRY_INTERVAL
}

/// `textStatus` (C++ 版は知らない値で abort する)
pub fn text_status(status: i32) -> Option<&'static [u8]> {
    match status {
        UNTRIED => Some(b"Untried"),
        SUCCESS => Some(b"Success"),
        ERROR => Some(b"Error"),
        _ => None,
    }
}

/// `UptestServiceRegistry::addURL` の判断。`valid` と `scheme` は `URI` の結果。
pub fn check_add_url(valid: bool, scheme: &[u8], url: &[u8], existing: &[&[u8]]) -> Result<(), &'static [u8]> {
    if !valid {
        Err(b"invalid URL")
    } else if scheme != b"http" {
        Err(b"unsupported protocol")
    } else if existing.contains(&url) {
        Err(b"URL already exists")
    } else {
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const DOC: &[u8] = b"<yp4g><yp name=\"YP\"/>\
        <host ip=\"192.168.0.1\" port_open=\"1\" speed=\"999\" over=\"0\"/>\
        <uptest checkable=\"1\" remain=\"0\"/>\
        <uptest_srv addr=\"example.com\" port=\"443\" object=\"/yp/uptest.cgi\" post_size=\"250\" limit=\"3000\" interval=\"15\" enabled=\"1\"/>\
        </yp4g>";

    #[test]
    fn read_info_ok() {
        let info = read_info(DOC).unwrap();
        let want: [&[u8]; 14] = [
            b"YP", b"192.168.0.1", b"1", b"999", b"0", b"1", b"0", b"example.com", b"443", b"/yp/uptest.cgi",
            b"250", b"3000", b"15", b"1",
        ];
        for (a, b) in info.iter().zip(want) {
            assert_eq!(a.as_slice(), b);
        }
        assert_eq!(post_url(&info[7], &info[8], &info[9]), b"http://example.com:443/yp/uptest.cgi");
    }

    #[test]
    fn read_info_quirks() {
        // findAttr は名前の先頭が一致すればよく、findNode は大文字小文字を問わない
        let doc = b"<YP nameX=\"a\"/><x><HOST ipv6=\"i\" port_openz=\"p\" speed=\"s\" over=\"o\"/></x>\
            <uptest checkable=\"c\" remain=\"r\"/><uptest_srv addr=\"a\" port_x=\"1\" object=\"\" post_size=\"2\" \
            limit=\"3\" interval=\"4\" enabled=\"5\"/>";
        // 根は最後の最上位のノード (uptest_srv) だけなので、yp は見つからない
        assert_eq!(read_info(doc), Err(ReadError::Null));
        let doc2 = [&b"<r>"[..], doc, b"</r>"].concat();
        let info = read_info(&doc2).unwrap();
        assert_eq!(info[0], b"a");
        assert_eq!(info[1], b"i");
        assert_eq!(info[8], b"1");
        assert_eq!(info[9], b"");
    }

    #[test]
    fn read_info_errors() {
        assert_eq!(read_info(b""), Err(ReadError::Null));
        assert_eq!(read_info(b"</a>"), Err(ReadError::Xml(xml::Error::UnexpectedEndTag)));
        assert_eq!(read_info(b"<?foo?>"), Err(ReadError::Xml(xml::Error::NotXml)));
        assert!(matches!(read_info(b"<a b=c>"), Err(ReadError::Attr(_))));
    }

    #[test]
    fn misc() {
        assert!(is_ready(UNTRIED, 0, 0));
        assert!(!is_ready(ERROR, 0, 0));
        assert!(!is_ready(SUCCESS, 100, 160));
        assert!(is_ready(SUCCESS, 100, 161));
        assert!(is_ready(ERROR, 100, 50)); // 符号なしの引き算
        assert_eq!(text_status(2), Some(&b"Error"[..]));
        assert_eq!(text_status(3), None);
        let ex: [&[u8]; 1] = [b"http://a/"];
        assert_eq!(check_add_url(false, b"http", b"x", &ex), Err(&b"invalid URL"[..]));
        assert_eq!(check_add_url(true, b"https", b"x", &ex), Err(&b"unsupported protocol"[..]));
        assert_eq!(check_add_url(true, b"http", b"http://a/", &ex), Err(&b"URL already exists"[..]));
        assert_eq!(check_add_url(true, b"http", b"http://b/", &ex), Ok(()));
    }
}
